text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+
+* publish its definition (with all configuration) via unique (per artefact type) artefact id
+  - publishes the artefact definition to the repository under the artefact id
+  - makes it resolvable via a Tremor URL
+* create an instance from the artefact referenced by its type and id (or via Tremor URL) with a unique (per artefact) instance id
+  - instantiates the artefact and publishes it to the registry under the artefact-id and instance-id
+  - makes it resolvable via a Tremor URL
+* link the instance to other compatible instances (e.g. pipeline -> offramp or onramp -> pipeline), thus creating a connected and actively running event flow
+
+There are also the following steps:
+
+* Disconnect an artefact instance from all its connections (e.g. disconnect an offramp from its pipelines)
+* Destroy an artefact instance and thus stop its operation
+* Remove the artefact definition from the repository
+
+The main scope of the Tremor Deployment language (Troy for short) are the lifecycle steps of deploying, connecting and running artefacts.
+It tries to be flexible and powerful by exposing statements for all lifecycle operations, while also optimizing for a convenient experience in the common case of deploying rather simple event flows.
+
+Troy files can be deployed to a Tremor instance or cluster on startup or via API endpoint. It is also possible to issue Troy statements inside a REPL-like setup, but this is outside of the scope of this RFC.
+
+Whenever a Troy file is deployed the statements within it are executed.
+
+Similar to trickle, Troy also supports modularization and `use` statements using the same mechanism as is employed in trickle (loading the "`used`" file contents into the current file). This allows for defining commonly used artefacts in separate files.
+
+## Example
+
+```
+
+# define a connector with arguments
+define http_server connector my_http_connector
+args
+  server = "Tremor/1.2.3",
+with
+  codec = "json",
+  interceptors = [ "newline" ],
+  config.headers = {
+      "Server": args.server
+  }
+end;
+
+# define a file connector
+define file connector my_file
+with
+    config.path = "/snot/badger.json",
+    codec = "json",
+    interceptors = [ "newline" ]
+end;
+
+# define the simplest pipeline
+define pipeline passthrough
+query
+    select event from in into out;
+end;
+
+# instantiate a connector
+create connector instance "01" from my_http_connector
+with
+    server = "Tremor/3.2.1"
+end;
+
+# define a flow - connecting artefact instances
+define flow my_flow
+links
+    # reference artefacts by id - these instances are not recreated if they are already created
+    connect "/connector/my_http_connector/01:out" to "/pipeline/passthrough:in";
+    # implicit default ports - here we auto-create instances
+    connect passthrough to my_file;
+end;
+
+# instantiate a flow - let the events flow ~~~
+create instance "01" from flow/my_flow;
+```
```

Comment:
We should capture a few more of the (non choosen) examples we discussed, it'll help explaining the alternatives and decisions behind the syntax and encurage discourse.

Reply:
added